### PR TITLE
Switch transformer for sequence classification

### PR DESCRIPTION
**Pull request by jlamprou**
# What does this PR do?
This adds a sequence classification head to the PyTorch implementation of SwitchTransformers, following the pattern of T5ForSequenceClassification since it is also an encoder-decoder sequence classification model.

# NOTE:
- [ ] Failing tests, because we haven't a Checkpoint trained on classification.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Hey @ArthurZucker and @younesbelkada I would greatly appreciate a review on this when you have a chance.

### CODE REVIEW

File: src/transformers/models/switch_transformers/modeling_switch_transformers.py [hunk 1]
```diff
@@ -127,6 +127,24 @@ def load_balancing_loss_func(router_probs: torch.Tensor, expert_indices: torch.T
...
+        self.dropout = nn.Dropout(p=0.1)
+        self.out_proj = nn.Linear(model_dim, config.num_labels, bias=False)
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        hidden_states = self.dropout(hidden_states)
+        hidden_states = self.dense(hidden_states)
+        hidden_states = torch.tanh(hidden_states)
+        hidden_states = self.dropout(hidden_states)
+        hidden_states = self.out_proj(hidden_states)
+        return hidden_states
```
Comment (younesbelkada) [inline src/transformers/models/switch_transformers/modeling_switch_transformers.py:145]:
```suggestion
# Copied from transformers.models.t5.modeling_t5.T5ClassificationHead with T5->SwitchTransformers
class SwtichTransformersClassificationHead(nn.Module):
    """Head for sentence-level classification tasks."""

    def __init__(self, config: SwitchTransformersConfig, model_dim: int):
        super().__init__()
        self.dense = nn.Linear(model_dim, model_dim, bias=False)
        self.dropout = nn.Dropout(p=0.1)
        self.out_proj = nn.Linear(model_dim, config.num_labels, bias=False)

    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
        hidden_states = self.dropout(hidden_states)
        hidden_states = self.dense(hidden_states)
        hidden_states = torch.tanh(hidden_states)
        hidden_states = self.dropout(hidden_states)
        hidden_states = self.out_proj(hidden_states)
        return hidden_states
```

Comment (younesbelkada) [reply src/transformers/models/switch_transformers/modeling_switch_transformers.py:145]: Let's add a copied from statement here as the module is a copy of `T5ClassificationHead`

File: src/transformers/models/switch_transformers/modeling_switch_transformers.py [hunk 2]
```diff
@@ -822,6 +840,13 @@ def _init_weights(self, module):
             module.shared.weight.data.normal_(mean=0.0, std=factor * 1.0)
             if hasattr(module, "lm_head") and not self.config.tie_word_embeddings:
                 module.lm_head.weight.data.normal_(mean=0.0, std=factor * 1.0)
+        elif isinstance(module, ClassificationHead):
```
Comment (younesbelkada) [inline src/transformers/models/switch_transformers/modeling_switch_transformers.py:843]:
```suggestion
        elif isinstance(module, SwitchTransformersClassificationHead):
```

File: src/transformers/models/switch_transformers/modeling_switch_transformers.py [hunk 3]
```diff
@@ -1866,3 +1891,249 @@ def forward(
...
+        decoder_config.is_encoder_decoder = False
+        self.decoder = SwitchTransformersStack(decoder_config, self.shared)
+        # Classifier head
+        self.classification_head = ClassificationHead(config, self.model_dim)
+
+        # Initialize weights and apply final processing
+        self.post_init()
+
+        # Model parallel
+        self.device_map = None
```
Comment (younesbelkada) [inline src/transformers/models/switch_transformers/modeling_switch_transformers.py:1928]:
```suggestion
```
This is not needed

File: src/transformers/models/switch_transformers/modeling_switch_transformers.py [hunk 4]
```diff
@@ -1866,3 +1891,249 @@ def forward(
...
+        encoder_config.use_cache = False
+        encoder_config.is_encoder_decoder = False
+        self.encoder = SwitchTransformersStack(encoder_config, self.shared)
+
+        decoder_config = copy.deepcopy(config)
+        decoder_config.is_decoder = True
+        decoder_config.is_encoder_decoder = False
+        self.decoder = SwitchTransformersStack(decoder_config, self.shared)
+        # Classifier head
+        self.classification_head = ClassificationHead(config, self.model_dim)
```
Comment (younesbelkada) [inline src/transformers/models/switch_transformers/modeling_switch_transformers.py:1922]:
```suggestion
        self.classification_head = SwitchTransformersClassificationHead(config, self.model_dim)
```

**Review by younesbelkada — COMMENTED** [review]
Looks clean already ! Thanks for your contribution ! Left few code improvements to be more in-line with our codebase - it seems there is also a failing CI test, would be happy to take a look before merging?

**Review by ArthurZucker — COMMENTED** [review]
Hey! Thanks for contributing. As there are no released checkpoints for sequence classification, we usually try to: 
1. Open an issue with the feature request 
2. If the issue has strong support from the community (usually around 10 likes for example) the add support for it 🤗 

In this case you are the first user to request this new class! 🤗

**Comment by jlamprou** [comment]
> Hey! Thanks for contributing. As there are no released checkpoints for sequence classification, we usually try to:
> 
>     1. Open an issue with the feature request
> 
>     2. If the issue has strong support from the community (usually around 10 likes for example) the add support for it 🤗
> 
> 
> In this case you are the first user to request this new class! 🤗

Hi, I'm working on Classification checkpoint, it will be available once the training is done.

**Comment by jlamprou** [comment]
@ArthurZucker @younesbelkada, I would also like your opinion about the usage of Z and Aux Loss on the Sequence Classification task. By my understanding Z loss is designed to improve the training stability of large language models and Mixture-of-Experts (MoE) models and the Aux loss for Multi-task learning and Regularization. ETA-50minutes for trained checkpoint on MNLI using the [run_glue.py](https://github.com/huggingface/transformers/blob/main/examples/pytorch/text-classification/run_glue.py).

**Comment by jlamprou** [comment]
A trained checkpoint on MNLI [here](https://huggingface.co/glamprou/switch-base-8-mnli)

**Comment by jlamprou** [comment]
@ArthurZucker @younesbelkada, Let's not let this go stale, all tests are passing and 2 checkpoints are available

File: tests/models/switch_transformers/test_modeling_switch_transformers.py [hunk 5]
```diff
@@ -582,6 +620,126 @@ def setUp(self):
         self.model_tester = SwitchTransformersModelTester(self)
         self.config_tester = ConfigTester(self, config_class=SwitchTransformersConfig, d_model=37)
 
+    def _create_and_check_torch_fx_tracing(self, config, inputs_dict, output_loss=False):
```
Comment (ArthurZucker) [inline tests/models/switch_transformers/test_modeling_switch_transformers.py:623]: if this is copied from, let's use copied form as well!

**Review by ArthurZucker — COMMENTED** [review]
Looks good to me, sorry for the late review. 
Let's just use copied from wherever possible (even in tests) and good to go

File: src/transformers/models/switch_transformers/modeling_switch_transformers.py [hunk 6]
```diff
@@ -1866,3 +1897,246 @@ def forward(
...
+        decoder_config.is_decoder = True
+        decoder_config.is_encoder_decoder = False
+        self.decoder = SwitchTransformersStack(decoder_config, self.shared)
+        # Classifier head
+        self.classification_head = SwitchTransformersClassificationHead(config)
+
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def get_input_embeddings(self):
```
[the remark this answers is not in the document]
Comment (jlamprou) [reply src/transformers/models/switch_transformers/modeling_switch_transformers.py:1933]: Its from the same file, SwitchTransformersForConditionalGeneration

**Comment by ArthurZucker** [comment]
feel free to ping me again for another review once Cis are green

**Comment by jlamprou** [comment]
> feel free to ping me again for another review once Cis are green

@ArthurZucker All tests are passing, no changes were needed, the test failing wasn't related to my commits. An empty commit to rerun tests was enough.

File: src/transformers/models/switch_transformers/modeling_switch_transformers.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/switch_transformers/modeling_switch_transformers.py:1933]: No I meant from `t5` this will be easier to maintain!

File: src/transformers/models/switch_transformers/modeling_switch_transformers.py [hunk 7]
```diff
@@ -1866,3 +1897,246 @@ def forward(
...
+    def set_input_embeddings(self, new_embeddings):
+        self.shared = new_embeddings
+        self.encoder.set_input_embeddings(new_embeddings)
+        self.decoder.set_input_embeddings(new_embeddings)
+
+    def _tie_weights(self):
+        if self.config.tie_word_embeddings:
+            self._tie_or_clone_weights(self.encoder.embed_tokens, self.shared)
+            self._tie_or_clone_weights(self.decoder.embed_tokens, self.shared)
+
```
Comment (ArthurZucker) [inline src/transformers/models/switch_transformers/modeling_switch_transformers.py:1945]: both of these should have the `copied from t5` mention!

File: src/transformers/models/switch_transformers/modeling_switch_transformers.py [hunk 8]
```diff
@@ -1866,3 +1897,246 @@ def forward(
...
+    def get_encoder(self):
+        return self.encoder
+
+    def get_decoder(self):
+        return self.decoder
+
+    def _prune_heads(self, heads_to_prune):
```
Comment (ArthurZucker) [inline src/transformers/models/switch_transformers/modeling_switch_transformers.py:1952]: here as well!

File: tests/models/switch_transformers/test_modeling_switch_transformers.py [hunk 9]
```diff
@@ -305,6 +315,26 @@ def create_and_check_decoder_model_past(
         # test that outputs are equal for slice
         self.parent.assertTrue(torch.allclose(output_from_past_slice, output_from_no_past_slice, atol=1e-3))
 
+    def create_and_check_with_sequence_classification_head(
```
Comment (ArthurZucker) [inline tests/models/switch_transformers/test_modeling_switch_transformers.py:318]: this needs a copied from T5 tests mention as well!

File: tests/models/switch_transformers/test_modeling_switch_transformers.py [hunk 10]
```diff
@@ -647,6 +806,39 @@ def test_decoder_model_past_with_3d_attn_mask(self):
             lm_labels,
         )
 
+    # SwitchTransformersForSequenceClassification does not support inputs_embeds
```
Comment (ArthurZucker) [inline tests/models/switch_transformers/test_modeling_switch_transformers.py:809]: this is also copied from T5 let's add the mention as well

**Review by ArthurZucker — COMMENTED** [review]
Thanks! Let's add more copied from, this will be easier to maintain. A few PRs changing the way we tie embeddings is coming so we will only have to apply changes from T5 😉
